GliaWidgets/View/Chat/Message/Content/CustomCard/WebMessageCardView.swift

```diff
@@ -0,0 +1,168 @@
+        let jsScript = WKUserScript(
+            source: .js,
+            injectionTime: .atDocumentEnd,
+            forMainFrameOnly: false
+        )
+        config.userContentController.addUserScript(jsScript)
+        config.userContentController.add(self, name: .gliaDomain)
+        return WKWebView(
+            frame: .zero,
+            configuration: config
+        )
+    }()
+
+    var delegate: WebMessageCardViewDelegate?
+
+    private let message: MessageRenderer.Message
+
+    init(message: MessageRenderer.Message, metadata: HTMLMetadata) {
+        self.message = message
+        super.init(frame: .zero)
+        setup(metadata: metadata)
+        layout()
+    }
+
+    @available(*, unavailable)
+    required init?(coder: NSCoder) {
+        fatalError("init(coder:) has not been implemented")
+    }
+
+    override var intrinsicContentSize: CGSize {
+        webView.scrollView.contentSize
+    }
+}
+
+// MARK: - Private
+
+private extension WebMessageCardView {
+    private func setup(metadata: HTMLMetadata) {
+        webView.loadHTMLString(metadata.html, baseURL: nil)
+        webView.navigationDelegate = self
+    }
+
+    private func layout() {
+        addSubview(webView)
+        webView.autoPinEdgesToSuperviewEdges()
+    }
+}
+
+// MARK: - WKScriptMessageHandler
+
+extension WebMessageCardView: WKScriptMessageHandler {
+    func userContentController(
+        _ userContentController: WKUserContentController,
+        didReceive message: WKScriptMessage
+    ) {
+        guard
+            message.name == .gliaDomain,
+            let body = message.body as? String
+        else { return }
+
+        let components = body.components(separatedBy: String.separator)
+
+        guard
+            let text = components.first,
+            let value = components.last
+        else { return }
+
+        delegate?.didSelectCustomCardOption(
+            self,
+            selectedOption: .init(
+                text: text,
+                value: value
+            ),
+            for: self.message.id
+        )
+    }
+}
+
+// MARK: - WKNavigationDelegate
+
+extension WebMessageCardView: WKNavigationDelegate {
+    func webView(
+        _ webView: WKWebView,
+        didFinish navigation: WKNavigation!
+    ) {
+        self.webView.evaluateJavaScript(.readyState) { [weak self] complete, _ in
+            guard
+                let self = self,
+                complete != nil
+            else { return }
+            self.invalidateIntrinsicContentSize()
+            self.delegate?.viewDidUpdateHight(self)
+        }
+    }
+
+    func webView(
+        _ webView: WKWebView,
+        decidePolicyFor navigationAction: WKNavigationAction,
+        decisionHandler: @escaping (WKNavigationActionPolicy) -> Void
+    ) {
+        guard let url = navigationAction.request.url else {
+            decisionHandler(.allow)
+            return
+        }
+
+        switch url.scheme {
```

Comment:
Maybe it would be beneficial to extract the logic inside as pure static function which gets url and returns the tuple of `(decisionHandlerResult: WKNavigationActionPolicy, delegateShouldSelectUrl: Bool )`? That will allow to easily unit test it.

Reply:
e67c22054de28ee2b5c206c3d4963777182655f1